src/main/java/dev/morling/onebrc/CalculateAverage_rby.java

```diff
@@ -0,0 +1,223 @@
+            var nextNL = position;
+            while (read-- > 0) {
+                if (bytes[read] == '\n') {
+                    nextNL += read;
+                    cuts[workers] = nextNL;
+                    break;
+                }
+            }
+            position -= chunk;
+            buf.rewind();
+        }
+        cuts[0] = 0L;
+        return cuts;
+
+    }
+
+    public static void main(String[] args) throws IOException {
+        var p = Paths.get(FILE);
+        var cpus = Runtime.getRuntime().availableProcessors();
+        final long[] cuts = cuts(p, cpus);
+
+        var stats = IntStream.range(0, cuts.length - 1)
+                .parallel()
+                .mapToObj((i) -> stats(p, cuts[i], cuts[i + 1]))
+                .reduce(Stats.IDENTITY, Stats::combine);
+
+        stats.print();
+
+    }
+
+    static record Stats(Map<String, Integer> indexes, int nextIx, int[] stats) {
+        private final static Stats IDENTITY = new Stats(new HashMap(), 0, new int[0]);
+        // not much optimization needed here
+        Stats combine(Stats other) {
+            if (this == IDENTITY) return other;
+            if (other == IDENTITY) return this;
+            var myNextIx = nextIx; 
+            for(var e : other.indexes.entrySet()) {
+                int ix;
+                var ixi = indexes.get(e.getKey());
+                if ( ixi == null) {
+                    ix = myNextIx++ * 4;
+                } else {
+                     ix = ixi.intValue() * 4;
+                }
+                var oix = e.getValue() * 4;
+                stats[ix] = Math.min(stats[ix], other.stats[oix]);
+                stats[ix + 1] = Math.max(stats[ix + 1], other.stats[oix + 1]);
+                stats[ix + 2] += other.stats[oix + 2];
+                stats[ix + 3] += other.stats[oix + 3];
+            }
+            return new Stats(indexes, myNextIx, stats);
+        }
+        // or here
+        void print() {
+            var iter = new TreeMap<>(indexes).entrySet().iterator();
+            System.out.print("{");
+            if (iter.hasNext()) {
+                var e = iter.next();
+                var ix = e.getValue().intValue() * 4;
+                var avg = Math.round(stats[ix + 2]/((double)stats[ix+3]))/10.0;
+                System.out.print(e.getKey() + "="
+                        + (stats[ix]/10.0) + "/"
+                        + avg + "/"
+                        + (stats[ix + 1]/10.0));
+            }
+            while(iter.hasNext()) {
+                var e = iter.next();
+                var ix = e.getValue().intValue() * 4;
+                var avg = Math.round(stats[ix + 2]/((double)stats[ix+3]))/10.0;
+                System.out.print(", " + e.getKey() + "="
+                        + (stats[ix]/10.0) + "/"
+                        + avg + "/"
+                        + (stats[ix + 1]/10.0)) ;
+            }
+            System.out.println("}");
+        }
+    }
+
+    static final int MAX_CITIES = 1000;
```

Comment:
Hum hum, there can be 10,000 station names. How comes that measurements-10000-unique-keys.txt is passing for this one? CC @AlexanderYastrebov

Reply:
Hi @gunnarmorling, thanks for checking this. In line 185, the array is expanded if there's an overflow.

Reply:
Ah yes, got it. Thanks!